<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Codon Usage Bias Analysis â€¢ cubar</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Codon Usage Bias Analysis">
<meta name="description" content="A suite of functions for rapid and flexible analysis of codon usage bias. It provides in-depth analysis at the codon level, including relative synonymous codon usage (RSCU), tRNA weight calculations, machine learning predictions for optimal or preferred codons, and visualization of codon-anticodon pairing. Additionally, it can calculate various gene- specific codon indices such as codon adaptation index (CAI), effective number of codons (ENC), fraction of optimal codons (Fop), tRNA adaptation index (tAI), mean codon stabilization coefficients (CSCg), and GC contents (GC/GC3s/GC4d). It also supports both standard and non-standard genetic code tables found in NCBI, as well as custom genetic code tables.">
<meta property="og:description" content="A suite of functions for rapid and flexible analysis of codon usage bias. It provides in-depth analysis at the codon level, including relative synonymous codon usage (RSCU), tRNA weight calculations, machine learning predictions for optimal or preferred codons, and visualization of codon-anticodon pairing. Additionally, it can calculate various gene- specific codon indices such as codon adaptation index (CAI), effective number of codons (ENC), fraction of optimal codons (Fop), tRNA adaptation index (tAI), mean codon stabilization coefficients (CSCg), and GC contents (GC/GC3s/GC4d). It also supports both standard and non-standard genetic code tables found in NCBI, as well as custom genetic code tables.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">cubar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/cubar.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/non_standard_genetic_code.html">Non-standard genetic code</a></li>
    <li><a class="dropdown-item" href="articles/theory.html">Theories behind cubar</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mt1022/cubar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="cubar">cubar<a class="anchor" aria-label="anchor" href="#cubar"></a>
</h1></div>
<!-- badges: start -->

<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>cubar is a package for codon usage bias analysis in R. Main features are as follows:</p>
<ul>
<li>Codon level analyses
<ul>
<li>Calculate tRNA weights;</li>
<li>Calculate relative synonymous codon usage (RSCU);</li>
<li>Machine learning-based inference of optimal codons;</li>
<li>Visualization codon-anticodon pairing relationships;</li>
</ul>
</li>
<li>Gene level analyses
<ul>
<li>Tabulate codon frequency of each coding sequence;</li>
<li>Measure codon usage similarity to highly expressed genes with Codon Adaptation Index (CAI);</li>
<li>Quantify the influnce of codon usage on mRNA stability with Mean Codon Stabilization Coefficients (CSCg);</li>
<li>Measure codon usage bias with the nonparametric index Effective number of codons (ENC);</li>
<li>Measure the fraction of pre-determined optimal codons (Fop) in each sequence;</li>
<li>Overall GC content (GC) or that of 3rd synonymous positions (GC3s) or 4-fold degenerate sites (GC4d);</li>
<li>Quantify whether codon usage matches tRNA availability using tRNA Adaptation Index (tAI);</li>
</ul>
</li>
<li>Utilities
<ul>
<li>Sliding window analysis of codon usage within a coding sequence;</li>
<li>Optimize codon usage based on optimal codons for heterologous expression;</li>
<li>Test differential usage of codons between two sets of sequences;</li>
</ul>
</li>
</ul>
<p>Main advantages of <code>cubar</code> are as follows: - Process large datasets (&gt;10,0000 sequences) efficiently using the <code>Biostrings</code> and <code>data.table</code> backends; - Support genetic codes cataloged by <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi" class="external-link">NCBI</a> as well as custom ones; - Integrate with other data analysis or bioinformatic packages in the R ecosystem;</p>
</div>
<div class="section level3">
<h3 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h3>
<p>Depends</p>
<ul>
<li>
<code>R</code> (&gt;= 4.1.0)</li>
</ul>
<p>Imports</p>
<ul>
<li>
<code>Biostrings</code> (&gt;= 2.60.0),</li>
<li>
<code>IRanges</code> (&gt;= 2.34.0),</li>
<li>
<code>data.table</code> (&gt;= 1.14.0),</li>
<li>
<code>ggplot2</code> (&gt;= 3.3.5),</li>
<li>
<code>rlang</code> (&gt;= 0.4.11)</li>
</ul>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>The latest release of <code>cubar</code> can be installed with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"cubar"</span><span class="op">)</span></span></code></pre></div>
<p>The latest developmental version of <code>cubar</code> can be installed with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"mt1022/cubar"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<p>Documentation can be found within R (by typing <code>?function_name</code>). The following tutorials are available from our <a href="https://mt1022.github.io/cubar/">website</a>:</p>
<ul>
<li>
<a href="https://mt1022.github.io/cubar/articles/cubar.html">Get Started</a>: A brief introduction demonstrating the basic usage of <code>cubar</code>;</li>
<li>
<a href="https://mt1022.github.io/cubar/articles/non_standard_genetic_code.html">Non-standard Genetic Code</a>: How to use <code>cubar</code> with non-standard genetic codes;</li>
<li>
<a href="https://mt1022.github.io/cubar/articles/theory.html">Theories behind cubar</a>: The mathematical details behind the core functions in <code>cubar</code>;</li>
</ul>
</div>
<div class="section level3">
<h3 id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h3>
<p>Please use GitHub <a href="https://github.com/mt1022/cubar/issues" class="external-link">issues</a> for bug reports, questions, and feature requests.</p>
</div>
<div class="section level3">
<h3 id="suggests">Suggests<a class="anchor" aria-label="anchor" href="#suggests"></a>
</h3>
<ul>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html" class="external-link">Biostrings</a> for sequence input/output and manipulation;</li>
<li>
<a href="https://github.com/dosorio/Peptides" class="external-link">Peptides</a> for peptide- or protein-related indices;</li>
</ul>
</div>
<div class="section level3">
<h3 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h3>
<p>GitHub Copilot was used to suggest code snippets in the development of this package. Thanks the <a href="https://education.github.com/" class="external-link">GitHub Education</a> teacher program for providing free access to GitHub Copilot.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=cubar" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/mt1022/cubar/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/mt1022/cubar/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing cubar</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Hong Zhang <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-4064-9432" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Bu Zi <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://CRAN.R-project.org/package=cubar" class="external-link"><img src="https://www.r-pkg.org/badges/version/cubar" alt="CRAN status"></a></li>
<li><a href="https://cran.r-project.org/package=cubar" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/cubar"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://doi.org/10.5281/zenodo.10155990" class="external-link"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.10155990.svg" alt="DOI"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hong Zhang, Bu Zi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
