<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Estimate tRNA weight w — est_trna_weight • cubar</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate tRNA weight w — est_trna_weight"><meta name="description" content="est_trna_weight compute the tRNA weight per codon for TAI calculation.
This weight reflects relative tRNA availability for each codon."><meta property="og:description" content="est_trna_weight compute the tRNA weight per codon for TAI calculation.
This weight reflects relative tRNA availability for each codon."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cubar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/cubar.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/non_standard_genetic_code.html">Non-standard genetic code</a></li>
    <li><a class="dropdown-item" href="../articles/theory.html">Theories behind cubar</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mt1022/cubar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Estimate tRNA weight w</h1>
      <small class="dont-index">Source: <a href="https://github.com/mt1022/cubar/blob/main/R/codon_weight.R" class="external-link"><code>R/codon_weight.R</code></a></small>
      <div class="d-none name"><code>est_trna_weight.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>est_trna_weight</code> compute the tRNA weight per codon for TAI calculation.
This weight reflects relative tRNA availability for each codon.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">est_trna_weight</span><span class="op">(</span></span>
<span>  <span class="va">trna_level</span>,</span>
<span>  codon_table <span class="op">=</span> <span class="fu"><a href="get_codon_table.html">get_codon_table</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  domain <span class="op">=</span> <span class="st">"Eukarya"</span>,</span>
<span>  s <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-trna-level-">trna_level, <a class="anchor" aria-label="anchor" href="#arg-trna-level-"></a></dt>
<dd><p>named vector of tRNA level (or gene copy numbers), one value for each anticodon.
vector names are anticodons.</p></dd>


<dt id="arg-codon-table">codon_table<a class="anchor" aria-label="anchor" href="#arg-codon-table"></a></dt>
<dd><p>a table of genetic code derived from <code>get_codon_table</code> or <code>create_codon_table</code>.</p></dd>


<dt id="arg-domain">domain<a class="anchor" aria-label="anchor" href="#arg-domain"></a></dt>
<dd><p>The taxonomic domain of interest. "Eukarya" (default), "Bacteria" or "Archaea".
Specify either the parameter "domain" or "s".</p></dd>


<dt id="arg-s">s<a class="anchor" aria-label="anchor" href="#arg-s"></a></dt>
<dd><p>list of non-Waston-Crick pairing panelty. Specify either the parameter "domain" or "s".</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>data.table of tRNA expression information.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>dos Reis M, Savva R, Wernisch L. 2004. Solving the riddle of codon usage preferences: a test for translational selection. Nucleic Acids Res 32:5036-5044.</p>
<p>Sabi R, Tuller T. 2014. Modelling the efficiency of codon-tRNA interactions based on codon usage bias. DNA Res 21:511-526.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># estimate codon tRNA weight for yeast</span></span></span>
<span class="r-in"><span><span class="va">yeast_trna_w</span> <span class="op">&lt;-</span> <span class="fu">est_trna_weight</span><span class="op">(</span><span class="va">yeast_trna_gcn</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">yeast_trna_w</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     aa_code amino_acid  codon subfam anticodon trna_id ac_level       W</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;char&gt;     &lt;char&gt; &lt;char&gt; &lt;char&gt;    &lt;char&gt;  &lt;char&gt;  &lt;table&gt;   &lt;num&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  1:       F        Phe    TTT Phe_TT       AAA Phe-AAA        0  2.1390</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2:       F        Phe    TTC Phe_TT       GAA Phe-GAA       10 10.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  3:       L        Leu    TTA Leu_TT       TAA Leu-TAA        7  7.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  4:       L        Leu    TTG Leu_TT       CAA Leu-CAA       10 12.5935</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  5:       S        Ser    TCT Ser_TC       AGA Ser-AGA       11 11.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  6:       S        Ser    TCC Ser_TC       GGA Ser-GGA        0  5.8751</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  7:       S        Ser    TCA Ser_TC       TGA Ser-TGA        3  4.0175</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  8:       S        Ser    TCG Ser_TC       CGA Ser-CGA        1  2.1115</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  9:       Y        Tyr    TAT Tyr_TA       ATA Tyr-ATA        0  1.7112</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10:       Y        Tyr    TAC Tyr_TA       GTA Tyr-GTA        8  8.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11:       C        Cys    TGT Cys_TG       ACA Cys-ACA        0  0.8556</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12:       C        Cys    TGC Cys_TG       GCA Cys-GCA        4  4.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13:       W        Trp    TGG Trp_TG       CCA Trp-CCA        6  6.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14:       L        Leu    CTT Leu_CT       AAG Leu-AAG        0  0.2139</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15:       L        Leu    CTC Leu_CT       GAG Leu-GAG        1  1.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16:       L        Leu    CTA Leu_CT       TAG Leu-TAG        3  3.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17:       L        Leu    CTG Leu_CT       CAG Leu-CAG        0  1.1115</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18:       P        Pro    CCT Pro_CC       AGG Pro-AGG        2  2.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19:       P        Pro    CCC Pro_CC       GGG Pro-GGG        0  1.0682</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20:       P        Pro    CCA Pro_CC       TGG Pro-TGG       10 10.1850</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21:       P        Pro    CCG Pro_CC       CGG Pro-CGG        0  3.7050</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 22:       H        His    CAT His_CA       ATG His-ATG        0  1.4973</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23:       H        His    CAC His_CA       GTG His-GTG        7  7.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24:       Q        Gln    CAA Gln_CA       TTG Gln-TTG        9  9.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25:       Q        Gln    CAG Gln_CA       CTG Gln-CTG        1  4.3345</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26:       R        Arg    CGT Arg_CG       ACG Arg-ACG        6  6.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27:       R        Arg    CGC Arg_CG       GCG Arg-GCG        0  3.2046</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28:       R        Arg    CGA Arg_CG       TCG Arg-TCG        0  0.5550</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 29:       R        Arg    CGG Arg_CG       CCG Arg-CCG        1  1.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 30:       I        Ile    ATT Ile_AT       AAT Ile-AAT       13 13.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 31:       I        Ile    ATC Ile_AT       GAT Ile-GAT        0  6.9433</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 32:       I        Ile    ATA Ile_AT       TAT Ile-TAT        2  3.2025</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 33:       M        Met    ATG Met_AT       CAT Met-CAT        5  5.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34:       T        Thr    ACT Thr_AC       AGT Thr-AGT       11 11.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35:       T        Thr    ACC Thr_AC       GGT Thr-GGT        0  5.8751</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 36:       T        Thr    ACA Thr_AC       TGT Thr-TGT        4  5.0175</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 37:       T        Thr    ACG Thr_AC       CGT Thr-CGT        1  2.4820</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 38:       N        Asn    AAT Asn_AA       ATT Asn-ATT        0  2.1390</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 39:       N        Asn    AAC Asn_AA       GTT Asn-GTT       10 10.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 40:       K        Lys    AAA Lys_AA       TTT Lys-TTT        7  7.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 41:       K        Lys    AAG Lys_AA       CTT Lys-CTT       14 16.5935</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 42:       S        Ser    AGT Ser_AG       ACT Ser-ACT        0  0.4278</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 43:       S        Ser    AGC Ser_AG       GCT Ser-GCT        2  2.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 44:       R        Arg    AGA Arg_AG       TCT Arg-TCT       11 11.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 45:       R        Arg    AGG Arg_AG       CCT Arg-CCT        1  5.0755</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 46:       V        Val    GTT Val_GT       AAC Val-AAC       14 14.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 47:       V        Val    GTC Val_GT       GAC Val-GAC        0  7.4774</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 48:       V        Val    GTA Val_GT       TAC Val-TAC        2  3.2950</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 49:       V        Val    GTG Val_GT       CAC Val-CAC        2  2.7410</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 50:       A        Ala    GCT Ala_GC       AGC Ala-AGC       11 11.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 51:       A        Ala    GCC Ala_GC       GGC Ala-GGC        0  5.8751</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 52:       A        Ala    GCA Ala_GC       TGC Ala-TGC        5  6.0175</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 53:       A        Ala    GCG Ala_GC       CGC Ala-CGC        0  1.8525</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 54:       D        Asp    GAT Asp_GA       ATC Asp-ATC        0  3.4224</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 55:       D        Asp    GAC Asp_GA       GTC Asp-GTC       16 16.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 56:       E        Glu    GAA Glu_GA       TTC Glu-TTC       14 14.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 57:       E        Glu    GAG Glu_GA       CTC Glu-CTC        2  7.1870</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 58:       G        Gly    GGT Gly_GG       ACC Gly-ACC        0  3.4224</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 59:       G        Gly    GGC Gly_GG       GCC Gly-GCC       16 16.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 60:       G        Gly    GGA Gly_GG       TCC Gly-TCC        3  3.0000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 61:       G        Gly    GGG Gly_GG       CCC Gly-CCC        2  3.1115</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     aa_code amino_acid  codon subfam anticodon trna_id ac_level       W</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              w</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;num&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  1: 0.12890590</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2: 0.60264561</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  3: 0.42185193</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  4: 0.75894175</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  5: 0.66291018</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  6: 0.35406032</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  7: 0.24211288</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  8: 0.12724862</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  9: 0.10312472</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10: 0.48211649</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11: 0.05156236</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12: 0.24105825</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13: 0.36158737</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14: 0.01289059</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15: 0.06026456</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16: 0.18079368</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17: 0.06698406</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18: 0.12052912</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19: 0.06437460</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20: 0.61379456</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21: 0.22328020</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 22: 0.09023413</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23: 0.42185193</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24: 0.54238105</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25: 0.26121674</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26: 0.36158737</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27: 0.19312381</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28: 0.03344683</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 29: 0.06026456</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 30: 0.78343930</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 31: 0.41843493</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 32: 0.19299726</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 33: 0.30132281</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34: 0.66291018</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35: 0.35406032</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 36: 0.30237744</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 37: 0.14957664</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 38: 0.12890590</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 39: 0.60264561</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 40: 0.42185193</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 41: 1.00000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 42: 0.02578118</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 43: 0.12052912</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 44: 0.66291018</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 45: 0.30587278</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 46: 0.84370386</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 47: 0.45062223</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 48: 0.19857173</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 49: 0.16518516</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 50: 0.66291018</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 51: 0.35406032</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 52: 0.36264200</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 53: 0.11164010</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 54: 0.20624944</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 55: 0.96423298</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 56: 0.84370386</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 57: 0.43312140</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 58: 0.20624944</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 59: 0.96423298</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 60: 0.18079368</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 61: 0.18751318</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              w</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hong Zhang, Mengyue Liu, Bu Zi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

