<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Codon optimization • cubar</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Codon optimization">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cubar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/cubar.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/codon_optimization.html">Codon optimization</a></li>
    <li><a class="dropdown-item" href="../articles/non_standard_genetic_code.html">Non-standard genetic code</a></li>
    <li><a class="dropdown-item" href="../articles/theory.html">Theories behind cubar</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mt1022/cubar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Codon optimization</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mt1022/cubar/blob/v1.2.0/vignettes/codon_optimization.Rmd" class="external-link"><code>vignettes/codon_optimization.Rmd</code></a></small>
      <div class="d-none name"><code>codon_optimization.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>Codon optimization modifies a gene’s coding sequence to enhance
protein production without changing the encoded amino acid sequence. The
<code>codon_optimize</code> function in the <code>cubar</code> package
provides three strategies for optimizing coding sequences based on the
codon usage of the target organism. The first two strategies replace
rare codons with more frequently used ones, while the third strategy
employs the third-party deep learning model
<code>CodonTransformer</code> (Fallahpour et al., 2025) to optimize
codon usage.</p>
</div>
<div class="section level3">
<h3 id="evironment-setup">Evironment setup<a class="anchor" aria-label="anchor" href="#evironment-setup"></a>
</h3>
<p>Additionally, <code>cubar</code> can integrate the state-of-the-art
deep learning model <code>SpliceAI</code> (Jaganathan et al., 2019) to
prevent the unintended introduction of cryptic splice sites in optimized
sequences. As both <code>SpliceAI</code> and
<code>CodonTransformer</code> are Python-based, users must manually
install these packages. Here we demonstrates how to install them using
<code>conda</code> (or <code>mamba</code>) in a new environment for use
in <code>cubar</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># create a new environment named "cubar_env" with both python and r installed</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> cubar_env python=3.12 r-base blas=<span class="pp">*</span>=netlib r-reticulate</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co"># activate the environment we just created</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">conda</span> activate cubar_env</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co"># install CodonTransformer and SpliceAI</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="ex">pip</span> install CodonTransformer tensorflow spliceai</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="optimization-strategies">Optimization strategies<a class="anchor" aria-label="anchor" href="#optimization-strategies"></a>
</h3>
<div class="section level5">
<h5 id="naive-method">“naive” method<a class="anchor" aria-label="anchor" href="#naive-method"></a>
</h5>
<p>The default “naive” method simply replaces each codon to the most
preferred one in the same family or subfamily.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mt1022/cubar" class="external-link">cubar</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">seq</span> <span class="op">&lt;-</span> <span class="st">'ATGCTACGA'</span></span>
<span><span class="va">cf_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_codons.html">count_codons</a></span><span class="op">(</span><span class="va">yeast_cds</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required namespace: Biostrings</span></span>
<span><span class="va">optimal_codons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/est_optimal_codons.html">est_optimal_codons</a></span><span class="op">(</span><span class="va">cf_all</span><span class="op">)</span></span>
<span><span class="va">seq_opt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codon_optimize.html">codon_optimize</a></span><span class="op">(</span><span class="va">seq</span>, <span class="va">optimal_codons</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seq_opt</span><span class="op">)</span></span>
<span><span class="co">#&gt; 9-letter DNAString object</span></span>
<span><span class="co">#&gt; seq: <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span></span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="idt-method">“IDT” method<a class="anchor" aria-label="anchor" href="#idt-method"></a>
</h5>
<p>The “IDT” option implements the method used by the codon optimization
tool of Integrated DNA Technologies. Briefly, this method randomly
selects synonymous codons from the same family or subfamily based on
their relative frequency, but excluding rare codons used below 10% in
the target organism.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_opt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codon_optimize.html">codon_optimize</a></span><span class="op">(</span><span class="va">seq</span>, cf <span class="op">=</span> <span class="va">cf_all</span>, method <span class="op">=</span> <span class="st">"IDT"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seq_opt</span><span class="op">)</span></span>
<span><span class="co">#&gt; 9-letter DNAString object</span></span>
<span><span class="co">#&gt; seq: <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span></span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="codontransformer-method">“CodonTransformer” method<a class="anchor" aria-label="anchor" href="#codontransformer-method"></a>
</h5>
<p>The “CodonTransformer” method optimizes codon usage with the
third-party software <code>CodonTransformer</code> directly using a
wrapper in R. <code>CodonTransformer</code> is a deep learning model
that can generate coding sequences that show similar codon usage and
distribution to host genes with reduced negative cis elements in a wide
range of organisms across the tree of life. Please refer to the original
study for more details.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_opt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codon_optimize.html">codon_optimize</a></span><span class="op">(</span><span class="va">seq</span>, method <span class="op">=</span> <span class="st">"CodonTransformer"</span>, organism <span class="op">=</span> <span class="st">"Saccharomyces cerevisiae"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seq_opt</span><span class="op">)</span></span>
<span><span class="co">#&gt; 9-letter DNAString object</span></span>
<span><span class="co">#&gt; seq: ATGTTAAGATGA</span></span></code></pre></div>
<p><code>cubar</code> can generate several optimized sequences at the
same time using the argument <code>num_sequences</code> with the method
“IDT” and “CodonTransformer”. When <code>num_sequences</code> is greater
than 1, identical duplicate sequences will be retained as a single copy,
potentially resulting in a final sequence count less than the specified
value.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqs_opt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codon_optimize.html">codon_optimize</a></span><span class="op">(</span><span class="va">seq</span>, cf <span class="op">=</span> <span class="va">cf_all</span>, method <span class="op">=</span> <span class="st">"IDT"</span>, num_sequences <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seqs_opt</span><span class="op">)</span></span>
<span><span class="co">#&gt; DNAStringSet object of length 6:</span></span>
<span><span class="co">#&gt;     width seq</span></span>
<span><span class="co">#&gt; [1]     9 ATGCTCCGT</span></span>
<span><span class="co">#&gt; [2]     9 ATGCTGCGT</span></span>
<span><span class="co">#&gt; [3]     9 ATGCTTCGT</span></span>
<span><span class="co">#&gt; [4]     9 ATGCTACGT</span></span>
<span><span class="co">#&gt; [5]     9 ATGCTCCGA</span></span>
<span><span class="co">#&gt; [6]     9 ATGCTTCGA</span></span>
<span><span class="va">seqs_opt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codon_optimize.html">codon_optimize</a></span><span class="op">(</span><span class="va">seq</span>, method <span class="op">=</span> <span class="st">"CodonTransformer"</span>, organism <span class="op">=</span> <span class="st">"Saccharomyces cerevisiae"</span>,</span>
<span>num_sequences <span class="op">=</span> <span class="fl">10</span>, deterministic <span class="op">=</span><span class="cn">FALSE</span>, temperature <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seqs_opt</span><span class="op">)</span></span>
<span><span class="co">#&gt; DNAStringSet object of length 4:</span></span>
<span><span class="co">#&gt;     width seq</span></span>
<span><span class="co">#&gt; [1]    12 ATGTTGAGATAA</span></span>
<span><span class="co">#&gt; [2]    12 ATGTTAAGATAA</span></span>
<span><span class="co">#&gt; [3]    12 ATGTTGAGATGA</span></span>
<span><span class="co">#&gt; [4]    12 ATGTTGAGATAG</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="splice-site-detection">Splice site detection<a class="anchor" aria-label="anchor" href="#splice-site-detection"></a>
</h3>
<p>In addition, <code>cubar</code> integrated the deep learning tool
<code>SpliceAI</code> to identify potential splice sites with the
argument <code>spliceai</code>. When the probabilities of non-splice
site for each base are greater than 0.5, it is considered that there are
no potential splice junction sites, and the
<code>Possible_splice_junction</code> in the output is marked as FALSE,
otherwise it is marked as TRUE.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqs_opt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codon_optimize.html">codon_optimize</a></span><span class="op">(</span><span class="va">seq</span>, cf <span class="op">=</span> <span class="va">cf_all</span>, method <span class="op">=</span> <span class="st">"IDT"</span>, num_sequences <span class="op">=</span> <span class="fl">10</span>, spliceai <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seqs_opt</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Candidate_optimized_sequence Possible_splice_junction</span></span>
<span><span class="co">#&gt;                           &lt;char&gt;                   &lt;lgcl&gt;</span></span>
<span><span class="co">#&gt;  1:                    ATGCTACGC                    FALSE</span></span>
<span><span class="co">#&gt;  2:                    ATGCTGCGA                    FALSE</span></span>
<span><span class="co">#&gt;  3:                    ATGCTGCGT                    FALSE</span></span>
<span><span class="co">#&gt;  4:                    ATGCTTCGC                    FALSE</span></span>
<span><span class="co">#&gt;  5:                    ATGCTACGT                    FALSE</span></span>
<span><span class="co">#&gt;  6:                    ATGCTTCGG                    FALSE</span></span>
<span><span class="co">#&gt;  7:                    ATGCTCCGT                    FALSE</span></span>
<span><span class="co">#&gt;  8:                    ATGCTTCGT                    FALSE</span></span>
<span><span class="co">#&gt;  9:                    ATGCTCCGA                    FALSE</span></span>
<span><span class="co">#&gt; 10:                    ATGCTTCGA                    FALSE</span></span>
<span><span class="va">seq_opt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codon_optimize.html">codon_optimize</a></span><span class="op">(</span><span class="va">seq</span>, method <span class="op">=</span> <span class="st">"CodonTransformer"</span>, organism <span class="op">=</span> <span class="st">"Saccharomyces cerevisiae"</span>, spliceai <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seq_opt</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Candidate_optimized_sequence Possible_splice_junction</span></span>
<span><span class="co">#&gt;                          &lt;char&gt;                   &lt;lgcl&gt;</span></span>
<span><span class="co">#&gt; 1:                 ATGTTAAGATGA                    FALSE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="recommendations">Recommendations<a class="anchor" aria-label="anchor" href="#recommendations"></a>
</h3>
<p>Codon usage within a coding sequence influences multiple aspects of
mRNA biology, including RNA secondary structure, translation elongation,
co-translational folding, and mRNA stability (Liu et al., 2021).
Achieving an optimal coding sequence requires carefully balancing all of
these factors, making codon optimization a more complex task than it may
initially appear. The <code>naive</code> approach implemented in
cubar—which simply replaces each non-optimal codon with the most
preferred one—fails to account for local sequence context and may lead
to unintended consequences. For instance, rare codons are often
strategically positioned to facilitate proper folding of protein
domains; indiscriminately replacing them with preferred codons can
disrupt this process and promote aggregation (Moss et al., 2024).</p>
<p>Given these complexities, we recommend using more sophisticated and
context-aware methods for codon optimization. Established strategies
such as that implemented in the IDT’s codon optimization tool have
demonstrated long-term effectiveness and are widely adopted within the
research community. Meanwhile, newer approaches based on deep generative
models—such as CodonTransformer (Fallahpour et al., 2025) and CodonBert
(Li et al., 2024)—leverage large-scale natural sequence data and
advanced architectures like recurrent neural networks and attention
mechanisms to capture context-dependent codon usage patterns (Novakovsky
et al., 2023). These deep learning–based models offer a powerful,
flexible framework for codon optimization and are likely to play a
central role in future sequence design workflows.</p>
<div class="section level5">
<h5 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h5>
<ul>
<li>Fallahpour A, Gureghian V, Filion GJ, Lindner AB, Pandi A.
CodonTransformer: a multispecies codon optimizer using context-aware
neural networks. Nat Commun. 2025 Apr 3;16(1):3205. doi:
10.1038/s41467-025-58588-7. PMID: 40180930; PMCID: PMC11968976.</li>
<li>Jaganathan K, Kyriazopoulou Panagiotopoulou S, McRae JF, Darbandi
SF, Knowles D, Li YI, Kosmicki JA, Arbelaez J, Cui W, Schwartz GB, Chow
ED, Kanterakis E, Gao H, Kia A, Batzoglou S, Sanders SJ, Farh KK.
Predicting Splicing from Primary Sequence with Deep Learning. Cell. 2019
Jan 24;176(3):535-548.e24. doi: 10.1016/j.cell.2018.12.015. Epub 2019
Jan 17. PMID: 30661751.</li>
<li>Method used by the IDT codon optimization tool: <a href="https://sg.idtdna.com/pages/education/decoded/article/idt-codon-optimization-tool-makes-synthetic-gene-design-easy" class="external-link uri">https://sg.idtdna.com/pages/education/decoded/article/idt-codon-optimization-tool-makes-synthetic-gene-design-easy</a>
</li>
<li>Liu Y, Yang Q, Zhao F. Synonymous but Not Silent: The Codon Usage
Code for Gene Expression and Protein Folding. Annu Rev Biochem. 2021 Jun
20;90:375-401. doi: 10.1146/annurev-biochem-071320-112701. Epub 2021 Jan
13. PMID: 33441035; PMCID: PMC8284178.</li>
<li>Moss MJ, Chamness LM, Clark PL. The Effects of Codon Usage on
Protein Structure and Folding. Annu Rev Biophys. 2024 Jul;53(1):87-108.
doi: 10.1146/annurev-biophys-030722-020555. Epub 2024 Jun 28. PMID:
38134335; PMCID: PMC11227313.</li>
<li>Li S, Moayedpour S, Li R, Bailey M, Riahi S, Kogler-Anele L, Miladi
M, Miner J, Pertuy F, Zheng D, Wang J, Balsubramani A, Tran K, Zacharia
M, Wu M, Gu X, Clinton R, Asquith C, Skaleski J, Boeglin L, Chivukula S,
Dias A, Strugnell T, Montoya FU, Agarwal V, Bar-Joseph Z, Jager S.
CodonBERT large language model for mRNA vaccines. Genome Res. 2024 Aug
20;34(7):1027-1035. doi: 10.1101/gr.278870.123. PMID: 38951026; PMCID:
PMC11368176.</li>
<li>Novakovsky G, Dexter N, Libbrecht MW, Wasserman WW, Mostafavi S.
Obtaining genetics insights from deep learning via explainable
artificial intelligence. Nat Rev Genet. 2023 Feb;24(2):125-137. doi:
10.1038/s41576-022-00532-2. Epub 2022 Oct 3. PMID: 36192604.</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hong Zhang, Mengyue Liu, Bu Zi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
