<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cubar">
<title>Theories behind cubar • cubar</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Theories behind cubar">
<meta property="og:description" content="cubar">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cubar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.2.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/cubar.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/theory.html">Theories behind cubar</a>
    <a class="dropdown-item" href="../articles/non_standard_genetic_code.html">Non-standard genetic code</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mt1022/cubar/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Theories behind cubar</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mt1022/cubar/blob/HEAD/vignettes/articles/theory.Rmd" class="external-link"><code>vignettes/articles/theory.Rmd</code></a></small>
      <div class="d-none name"><code>theory.Rmd</code></div>
    </div>

    
    
<div class="section level4">
<h4 id="the-codon-adaptation-index-cai">The Codon Adaptation Index (CAI)<a class="anchor" aria-label="anchor" href="#the-codon-adaptation-index-cai"></a>
</h4>
<p>CAI measures the similarity of codon usage between a coding sequence
(CDS) and a set of highly expressed genes (Sharp and Li 1987). To
quantify the relative synonymous codon usage (RSCU) values among the set
of highly expressed genes, the observed frequency of each codon is
divided by the frequency expected under the assumption of equal usage of
the synonymous codons for an amino acid (referred to as codon family
hereafter):</p>
<p><span class="math display">\[
\text{RSCU}_{ij} =
\frac{X_{ij}}{\frac{1}{n_i}\sum_{k=1}^{n_i}{X_{ik}}}=\frac{X_{ij}}{\frac{1}{n_i}X_i}
\]</span></p>
<p>where <span class="math inline">\(X_{ij}\)</span> is the number of
occurrences of the <span class="math inline">\(j\)</span>th codon for
the <span class="math inline">\(i\)</span>th codon family in CDSs of
highly expressed genes, and <span class="math inline">\(n_i\)</span> is
the total number of alternative codons for the <span class="math inline">\(i\)</span>th codon family. <span class="math inline">\(X_i\)</span> is the number of occurrence of codons
in the <span class="math inline">\(i\)</span>th codon family. The
relative adaptiveness of a codon, <span class="math inline">\(w\)</span>, is defined as the RSCU of that codon
divided by the maximum RSCU of that codon family:</p>
<p><span class="math display">\[
w_{ij}=\frac{\text{RSCU}_{ij}}{\max_{k=1,\dots,n_i}{\text{RSCU}_{ik}}}
\]</span></p>
<p>The CAI for a CDS is then calculated as the geometric mean of the
relative adaptiveness of codons used in the CDS of that gene:</p>
<p><span class="math display">\[
\text{CAI} = \bigg(\prod_{k=1}^L w_k\bigg)^{\frac{1}{L}}
=\text{exp}\bigg(\frac{1}{L}\sum_{k=1}^L \ln w_k\bigg) =
\text{exp}\bigg( \frac{\sum_i\sum_j {X_{ij}\ln w_{ij}}}{\sum_i\sum_j
X_{ij}} \bigg)
\]</span></p>
<p>where <span class="math inline">\(L\)</span> is the total number of
codons in CDS.</p>
</div>
<div class="section level4">
<h4 id="the-effective-number-of-codons-enc">The effective number of codons (ENC)<a class="anchor" aria-label="anchor" href="#the-effective-number-of-codons-enc"></a>
</h4>
<p>ENC reflects the unequal usage of codons in the CDS of a gene (Wright
1990). The lower the ENC, the larger the overall bias in codon usage.
The original implementation of ENC calculates the homozygosity of codon
usage in codon family <span class="math inline">\(i\)</span> as
follows:</p>
<p><span class="math display">\[
F_i^{\text{ori}} = \frac{X_i \sum_{j=1}^{n_i}{(\frac{X_{ij}}{X_i})^2} -
1}{X_i - 1}
\]</span></p>
<p>where <span class="math inline">\(X_i\)</span> is the number of
occurrence of codons in the <span class="math inline">\(i\)</span>th
codon family as used above. In cubar, codon family homozygosity is
calculated with an improved implementation that is more robust to bias
due to small <span class="math inline">\(n_i\)</span> (Sun, et
al. 2013):</p>
<p><span class="math display">\[
F_i = \sum_{j=1}^{n_i}(\frac{X_{ij} + 1}{X_i + n_i})^2
\]</span></p>
<p>Then we can calculate the final ENC of the CDS of a gene as
follows:</p>
<p><span class="math display">\[
\text{ENC}=K_1 +
K_2\frac{\sum_{i=1}^{K_2}{X_i}}{\sum_{i=1}^{K_2}{X_iF_i}} +
K_3\frac{\sum_{i=1}^{K_3}{X_i}}{\sum_{i=1}^{K_3}{X_iF_i}} +
K_4\frac{\sum_{i=1}^{K_4}{X_i}}{\sum_{i=1}^{K_4}{X_iF_i}}
\]</span></p>
<p>where <span class="math inline">\(K_m\)</span> denotes the number of
codon families with <span class="math inline">\(m\)</span> synonymous
codons:</p>
<p><span class="math display">\[
K_m=\sum_{i}{\delta(n_i-m)}
\]</span></p>
<p>It should be noted that throughout cubar, codon families with more
than four synomymous codons are divided into different subfamilies based
on the first two nucleotides of codons.</p>
</div>
<div class="section level4">
<h4 id="the-fraction-of-optimal-codons-f_op">The fraction of optimal codons (<span class="math inline">\(F_{op}\)</span>)<a class="anchor" aria-label="anchor" href="#the-fraction-of-optimal-codons-f_op"></a>
</h4>
<p><span class="math inline">\(F_{op}\)</span> measures the fraction of
optimal codons in the CDS of a gene given a list of optimal codons
(Ikemura 1981). It is calculated as follows:</p>
<p><span class="math display">\[
F_{\text{op}}=\frac{\sum_{k=1}^L{I(k\text{-th codon is optimal})}}{L}
\]</span></p>
<p>where <span class="math inline">\(I\)</span> is the indicator
function. In case the optimal codons are unknown, cubar automatically
determines the optimal codons in each codon family based on the
rationale that optimal codons tend to be used more frequently in genes
showing stronger codon usage bias. Specifically, as the number of
occurrence of codon <span class="math inline">\(j\)</span> among codon
family <span class="math inline">\(i\)</span> in gene <span class="math inline">\(k\)</span> follows a Binomial distribution:</p>
<p><span class="math display">\[
X_{ij}^k \sim \text{Binomial}(X_i^k, p_k)
\]</span></p>
<p>the influence of codon usage bias on the tendency to use codon <span class="math inline">\(j\)</span> can be estimated with binomial
regression using the following link function:</p>
<p><span class="math display">\[
\ln{\frac{p}{1-p}} \sim \beta_0 + \beta_1 \cdot \text{ENC} + \epsilon
\]</span></p>
<p>If codon <span class="math inline">\(j\)</span> is more likely to be
used in genes with higher overall codon bias (i.e., smaller ENC), then
the regression coefficient should be negative and significantly differs
from zero. cubar implements the binomial regression using the
<code>glm</code> function with <code>binomial</code> family in R.
Optimal codons are determined with a false discovery rate of 0.001 after
multiple testing correction with the Benjamini-Hochberg procedure
(Benjamini and Hochberg 1995) by default.</p>
</div>
<div class="section level4">
<h4 id="trna-adaptation-index-tai">tRNA Adaptation Index (tAI)<a class="anchor" aria-label="anchor" href="#trna-adaptation-index-tai"></a>
</h4>
<p>tAI quantifies how much the usage of codons in CDS of a gene
resembles the abundance of tRNAs (dos Reis, et al. 2004), which is often
approximated by tRNA gene copy numbers. To determine tAI, the absolute
tRNA adaptiveness value <span class="math inline">\(W_i\)</span> for
each codon <span class="math inline">\(i\)</span> is defined as</p>
<p><span class="math display">\[
W_i = \sum_{j=1}^{t_i}{(1 - s_{ij}) T_{ij}}
\]</span></p>
<p>where <span class="math inline">\(t_i\)</span> is the number of tRNA
isoacceptors recognizing the <span class="math inline">\(i\)</span>th
codon and <span class="math inline">\(T_i\)</span> is the abundance or
gene copy number of the <span class="math inline">\(j\)</span>th tRNA
recognizing this codon. <span class="math inline">\(s_{ij}\)</span> is a
panelty on non-canonical codon–anticodon pairings and differs among
different species (Sabi and Tuller 2014). Cubar uses average <span class="math inline">\(s_{ij}\)</span> values for eukaryotes (Sabi and
Tuller 2014) by default. Absolute adaptiveness values are then
normalized by maximum as follows:</p>
<p><span class="math display">\[
w_i =  \begin{cases} \frac{W_i}{\max_{j}W_j},&amp; \text{if } W_i &gt; 0
\\ \frac{\bar W|_{W_j \neq 0}}{\max_j{W_j}}, &amp; \text{if }
W_i=0\end{cases}
\]</span></p>
<p>where <span class="math inline">\(\bar W|_{W_j \neq 0}\)</span> is
the geometric mean of non-zero absolute adaptiveness values. Then tAI of
codons in a gene can be calculated as follows, similar to CAI:</p>
<p><span class="math display">\[
\text{TAI} = \bigg(\prod_{k=1}^L w_k\bigg)^{\frac{1}{L}}
=\text{exp}\bigg(\frac{1}{L}\sum_{k=1}^L \ln w_k\bigg) =
\text{exp}\bigg( \frac{\sum_i\sum_j {X_{ij}\ln w_{ij}}}{\sum_i\sum_j
X_{ij}} \bigg)
\]</span></p>
</div>
<div class="section level4">
<h4 id="the-mean-codon-stabilization-coefficients-cscg">The mean codon stabilization coefficients (CSCg)<a class="anchor" aria-label="anchor" href="#the-mean-codon-stabilization-coefficients-cscg"></a>
</h4>
<p>CSC of a codon is the Pearson correlation coefficient between the
frequency of that codon and the mRNA half-lives across different genes
(Presnyak, et al. 2015). CSCg is the average codon stabilization
coefficient (CSC) of codons in the CDS of a gene (Carneiro, et
al. 2019):</p>
<p><span class="math display">\[
\text{CSCg} = \frac{1}{L}\sum_{k=1}^L{\text{CSC}_k}= \frac{\sum_i\sum_j
{X_{ij} \text{CSC}_{ij}}}{\sum_i{\sum_j{X_{ij}}}}
\]</span></p>
</div>
<div class="section level4">
<h4 id="the-deviation-from-proportionality-d_p">The deviation from proportionality (<span class="math inline">\(D_p\)</span>)<a class="anchor" aria-label="anchor" href="#the-deviation-from-proportionality-d_p"></a>
</h4>
<p><span class="math inline">\(D_p\)</span> measures the departure of
the codon usage of an exogenous CDS from the tRNA pool of the host
organism (Chen, et al. 2020; Chen and Yang 2022). For a codon family
<span class="math inline">\(i\)</span> with <span class="math inline">\(n_i\)</span> synonymous codons (<span class="math inline">\(n_i&gt;1\)</span>), the fraction of codon <span class="math inline">\(j\)</span> among all occurrences of this family in
the exogenous CDS is:</p>
<p><span class="math display">\[
Y_{ij} = \frac{X_{ij}}{X_i}
\]</span></p>
<p>The relative tRNA availability of this codon in the codon family in
the host organism is calculated as:</p>
<p><span class="math display">\[
R_{ij} = \frac{w_{ij}}{\sum_{i=1}^{n_i}{w_{ij}}}
\]</span></p>
<p>where <span class="math inline">\(w_{ij}\)</span> can be relative
tRNA adaptiveness values as in the calculation of tAI or the RSCU of
host protein-coding genes as in the calculation of CAI. The discrepancy
between codon proportions of exogenous CDS and host tRNA availability is
then calculated as the Euclidean distance:</p>
<p><span class="math display">\[
D_i = \sqrt{\sum(Y_{ij} - R_{ij})^2}
\]</span></p>
<p>Finally, <span class="math inline">\(D_p\)</span> is calculated as
the geometric mean of distances for all codon families with at least two
synonymous codons:</p>
<p><span class="math display">\[
D_p = (\prod_{i=1}^K{D_i})^{\frac{1}{K}} = \exp
(\frac{1}{K}\sum_{i=1}^K{\ln D_i})
\]</span></p>
<p>where <span class="math inline">\(K=\sum_{i}{I(n_i&gt;1)}\)</span>
and <span class="math inline">\(I\)</span> is the indicator
function.</p>
</div>
<div class="section level4">
<h4 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h4>
<ul>
<li>Benjamini Y, Hochberg Y. 1995. Controlling the False Discovery Rate:
A Practical and Powerful Approach to Multiple Testing. Journal of the
Royal Statistical Society: Series B (Methodological) 57:289-300.</li>
<li>Carneiro RL, Requião RD, Rossetto S, Domitrovic T, Palhano FL. 2019.
Codon stabilization coefficient as a metric to gain insights into mRNA
stability and codon bias and their relationships with translation.
Nucleic Acids Res 47:2216-2228.</li>
<li>Chen F, Wu P, Deng S, Zhang H, Hou Y, Hu Z, Zhang J, Chen X, Yang
JR. 2020. Dissimilation of synonymous codon usage bias in virus-host
coevolution due to translational selection. Nat Ecol Evol
4:589-600.</li>
<li>Chen F, Yang JR. 2022. Distinct codon usage bias evolutionary
patterns between weakly and strongly virulent respiratory viruses.
iScience 25:103682.</li>
<li>dos Reis M, Savva R, Wernisch L. 2004. Solving the riddle of codon
usage preferences: a test for translational selection. Nucleic Acids Res
32:5036-5044.</li>
<li>Ikemura T. 1981. Correlation between the abundance of Escherichia
coli transfer RNAs and the occurrence of the respective codons in its
protein genes: a proposal for a synonymous codon choice that is optimal
for the E. coli translational system. J Mol Biol 151:389-409.</li>
<li>Presnyak V, Alhusaini N, Chen YH, Martin S, Morris N, Kline N, Olson
S, Weinberg D, Baker KE, Graveley BR, et al. 2015. Codon optimality is a
major determinant of mRNA stability. Cell 160:1111-1124.</li>
<li>Sabi R, Tuller T. 2014. Modelling the efficiency of codon-tRNA
interactions based on codon usage bias. DNA Res 21:511-526.</li>
<li>Sharp PM, Li WH. 1987. The codon Adaptation Index–a measure of
directional synonymous codon usage bias, and its potential applications.
Nucleic Acids Res 15:1281-1295.</li>
<li>Sun X, Yang Q, Xia X. 2013. An improved implementation of effective
number of codons (nc). Mol Biol Evol 30:191-196.</li>
<li>Wright F. 1990. The ‘effective number of codons’ used in a gene.
Gene 87:23-29.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Hong Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
