<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Theories behind cubar • cubar</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Theories behind cubar">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cubar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/cubar.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/codon_optimization.html">Codon optimization</a></li>
    <li><a class="dropdown-item" href="../articles/non_standard_genetic_code.html">Non-standard genetic code</a></li>
    <li><a class="dropdown-item" href="../articles/theory.html">Theories behind cubar</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mt1022/cubar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Theories behind cubar</h1>
                        <h4 data-toc-skip class="author">Bu Zi (draft),
Hong Zhang (revise)</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mt1022/cubar/blob/main/vignettes/articles/theory.Rmd" class="external-link"><code>vignettes/articles/theory.Rmd</code></a></small>
      <div class="d-none name"><code>theory.Rmd</code></div>
    </div>

    
    
<div class="section level4">
<h4 id="the-codon-adaptation-index-cai">The Codon Adaptation Index (CAI)<a class="anchor" aria-label="anchor" href="#the-codon-adaptation-index-cai"></a>
</h4>
<p>CAI measures the similarity of codon usage between a coding sequence
(CDS) and a set of highly expressed genes (Sharp and Li 1987). To
quantify the relative synonymous codon usage (RSCU) values among the set
of highly expressed genes, the observed frequency of each codon is
divided by the frequency expected under the assumption of equal usage of
the synonymous codons for an amino acid (referred to as codon family
hereafter):</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mtext mathvariant="normal">RSCU</mtext><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mfrac><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><mfrac><mn>1</mn><msub><mi>n</mi><mi>i</mi></msub></mfrac><munderover><mo>∑</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>n</mi><mi>i</mi></msub></munderover><msub><mi>X</mi><mrow><mi>i</mi><mi>k</mi></mrow></msub></mrow></mfrac><mo>=</mo><mfrac><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><mfrac><mn>1</mn><msub><mi>n</mi><mi>i</mi></msub></mfrac><msub><mi>X</mi><mi>i</mi></msub></mrow></mfrac></mrow><annotation encoding="application/x-tex">
\text{RSCU}_{ij} = \frac{X_{ij}}{\frac{1}{n_i}\sum_{k=1}^{n_i}{X_{ik}}}=\frac{X_{ij}}{\frac{1}{n_i}X_i}
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">X_{ij}</annotation></semantics></math>
is the number of occurrences of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>th
codon for the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>th
codon family in CDSs of highly expressed genes, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>n</mi><mi>i</mi></msub><annotation encoding="application/x-tex">n_i</annotation></semantics></math>
is the total number of alternative codons for the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>th
codon family.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>X</mi><mi>i</mi></msub><annotation encoding="application/x-tex">X_i</annotation></semantics></math>
is the number of occurrence of codons in the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>th
codon family. The relative adaptiveness of a codon,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>,
is defined as the RSCU of that codon divided by the maximum RSCU of that
codon family:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mfrac><msub><mtext mathvariant="normal">RSCU</mtext><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><munder><mo>max</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><msub><mi>n</mi><mi>i</mi></msub></mrow></munder><msub><mtext mathvariant="normal">RSCU</mtext><mrow><mi>i</mi><mi>k</mi></mrow></msub></mrow></mfrac></mrow><annotation encoding="application/x-tex">
w_{ij}=\frac{\text{RSCU}_{ij}}{\max_{k=1,\dots,n_i}{\text{RSCU}_{ik}}}
</annotation></semantics></math></p>
<p>The CAI for a CDS is then calculated as the geometric mean of the
relative adaptiveness of codons used in the CDS of that gene:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">CAI</mtext><mo>=</mo><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><munderover><mo>∏</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>L</mi></munderover><msub><mi>w</mi><mi>k</mi></msub><msup><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo><mfrac><mn>1</mn><mi>L</mi></mfrac></msup><mo>=</mo><mtext mathvariant="normal">exp</mtext><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><mfrac><mn>1</mn><mi>L</mi></mfrac><munderover><mo>∑</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>L</mi></munderover><mo>ln</mo><msub><mi>w</mi><mi>k</mi></msub><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo><mo>=</mo><mtext mathvariant="normal">exp</mtext><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><mfrac><mrow><munder><mo>∑</mo><mi>i</mi></munder><munder><mo>∑</mo><mi>j</mi></munder><mrow><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>ln</mo><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mrow><mrow><munder><mo>∑</mo><mi>i</mi></munder><munder><mo>∑</mo><mi>j</mi></munder><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mfrac><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">
\text{CAI} = \bigg(\prod_{k=1}^L w_k\bigg)^{\frac{1}{L}} =\text{exp}\bigg(\frac{1}{L}\sum_{k=1}^L \ln w_k\bigg) = \text{exp}\bigg( \frac{\sum_i\sum_j {X_{ij}\ln w_{ij}}}{\sum_i\sum_j X_{ij}} \bigg)
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>L</mi><annotation encoding="application/x-tex">L</annotation></semantics></math>
is the total number of codons in CDS.</p>
</div>
<div class="section level4">
<h4 id="the-effective-number-of-codons-enc">The effective number of codons (ENC)<a class="anchor" aria-label="anchor" href="#the-effective-number-of-codons-enc"></a>
</h4>
<p>ENC reflects the unequal usage of codons in the CDS of a gene (Wright
1990). The lower the ENC, the larger the overall bias in codon usage.
The original implementation of ENC calculates the homozygosity of codon
usage in codon family
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
as follows:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>F</mi><mi>i</mi><mtext mathvariant="normal">ori</mtext></msubsup><mo>=</mo><mfrac><mrow><msub><mi>X</mi><mi>i</mi></msub><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>n</mi><mi>i</mi></msub></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><mfrac><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mi>X</mi><mi>i</mi></msub></mfrac><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup><mo>−</mo><mn>1</mn></mrow><mrow><msub><mi>X</mi><mi>i</mi></msub><mo>−</mo><mn>1</mn></mrow></mfrac></mrow><annotation encoding="application/x-tex">
F_i^{\text{ori}} = \frac{X_i \sum_{j=1}^{n_i}{(\frac{X_{ij}}{X_i})^2} - 1}{X_i - 1}
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>X</mi><mi>i</mi></msub><annotation encoding="application/x-tex">X_i</annotation></semantics></math>
is the number of occurrence of codons in the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>th
codon family as used above. In cubar, codon family homozygosity is
calculated with an improved implementation that is more robust to bias
due to small
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>n</mi><mi>i</mi></msub><annotation encoding="application/x-tex">n_i</annotation></semantics></math>
(Sun, et al. 2013):</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>F</mi><mi>i</mi></msub><mo>=</mo><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>n</mi><mi>i</mi></msub></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><mfrac><mrow><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>+</mo><mn>1</mn></mrow><mrow><msub><mi>X</mi><mi>i</mi></msub><mo>+</mo><msub><mi>n</mi><mi>i</mi></msub></mrow></mfrac><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">
F_i = \sum_{j=1}^{n_i}(\frac{X_{ij} + 1}{X_i + n_i})^2
</annotation></semantics></math></p>
<p>Then we can calculate the final ENC of the CDS of a gene as
follows:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">ENC</mtext><mo>=</mo><msub><mi>K</mi><mn>1</mn></msub><mo>+</mo><msub><mi>K</mi><mn>2</mn></msub><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>K</mi><mn>2</mn></msub></munderover><msub><mi>X</mi><mi>i</mi></msub></mrow><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>K</mi><mn>2</mn></msub></munderover><mrow><msub><mi>X</mi><mi>i</mi></msub><msub><mi>F</mi><mi>i</mi></msub></mrow></mrow></mfrac><mo>+</mo><msub><mi>K</mi><mn>3</mn></msub><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>K</mi><mn>3</mn></msub></munderover><msub><mi>X</mi><mi>i</mi></msub></mrow><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>K</mi><mn>3</mn></msub></munderover><mrow><msub><mi>X</mi><mi>i</mi></msub><msub><mi>F</mi><mi>i</mi></msub></mrow></mrow></mfrac><mo>+</mo><msub><mi>K</mi><mn>4</mn></msub><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>K</mi><mn>4</mn></msub></munderover><msub><mi>X</mi><mi>i</mi></msub></mrow><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>K</mi><mn>4</mn></msub></munderover><mrow><msub><mi>X</mi><mi>i</mi></msub><msub><mi>F</mi><mi>i</mi></msub></mrow></mrow></mfrac></mrow><annotation encoding="application/x-tex">
\text{ENC}=K_1 + K_2\frac{\sum_{i=1}^{K_2}{X_i}}{\sum_{i=1}^{K_2}{X_iF_i}} + K_3\frac{\sum_{i=1}^{K_3}{X_i}}{\sum_{i=1}^{K_3}{X_iF_i}} + K_4\frac{\sum_{i=1}^{K_4}{X_i}}{\sum_{i=1}^{K_4}{X_iF_i}}
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>K</mi><mi>m</mi></msub><annotation encoding="application/x-tex">K_m</annotation></semantics></math>
denotes the number of codon families with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
synonymous codons:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>m</mi></msub><mo>=</mo><munder><mo>∑</mo><mi>i</mi></munder><mrow><mi>δ</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>n</mi><mi>i</mi></msub><mo>−</mo><mi>m</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mrow><annotation encoding="application/x-tex">
K_m=\sum_{i}{\delta(n_i-m)}
</annotation></semantics></math></p>
<p>It should be noted that throughout cubar, codon families with more
than four synomymous codons are divided into different subfamilies based
on the first two nucleotides of codons.</p>
</div>
<div class="section level4">
<h4 id="the-fraction-of-optimal-codons-fop">The fraction of optimal codons (Fop)<a class="anchor" aria-label="anchor" href="#the-fraction-of-optimal-codons-fop"></a>
</h4>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>F</mi><mrow><mi>o</mi><mi>p</mi></mrow></msub><annotation encoding="application/x-tex">F_{op}</annotation></semantics></math>
measures the fraction of optimal codons in the CDS of a gene given a
list of optimal codons (Ikemura 1981). It is calculated as follows:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>F</mi><mtext mathvariant="normal">op</mtext></msub><mo>=</mo><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>L</mi></munderover><mrow><mi>I</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>k</mi><mtext mathvariant="normal">-th codon is optimal</mtext><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mrow><mi>L</mi></mfrac></mrow><annotation encoding="application/x-tex">
F_{\text{op}}=\frac{\sum_{k=1}^L{I(k\text{-th codon is optimal})}}{L}
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>I</mi><annotation encoding="application/x-tex">I</annotation></semantics></math>
is the indicator function. In case the optimal codons are unknown, cubar
automatically determines the optimal codons in each codon family based
on the rationale that optimal codons tend to be used more frequently in
genes showing stronger codon usage bias. Specifically, as the number of
occurrence of codon
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
among codon family
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
in gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
follows a Binomial distribution:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow><mi>k</mi></msubsup><mo>∼</mo><mtext mathvariant="normal">Binomial</mtext><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>X</mi><mi>i</mi><mi>k</mi></msubsup><mo>,</mo><msub><mi>p</mi><mi>k</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
X_{ij}^k \sim \text{Binomial}(X_i^k, p_k)
</annotation></semantics></math></p>
<p>the influence of codon usage bias on the tendency to use codon
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
can be estimated with binomial regression using the following link
function:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>ln</mo><mfrac><mi>p</mi><mrow><mn>1</mn><mo>−</mo><mi>p</mi></mrow></mfrac><mo>∼</mo><msub><mi>β</mi><mn>0</mn></msub><mo>+</mo><msub><mi>β</mi><mn>1</mn></msub><mo>⋅</mo><mtext mathvariant="normal">ENC</mtext><mo>+</mo><mi>ϵ</mi></mrow><annotation encoding="application/x-tex">
\ln{\frac{p}{1-p}} \sim \beta_0 + \beta_1 \cdot \text{ENC} + \epsilon
</annotation></semantics></math></p>
<p>If codon
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
is more likely to be used in genes with higher overall codon bias (i.e.,
smaller ENC), then the regression coefficient should be negative and
significantly differs from zero. cubar implements the binomial
regression using the <code>glm</code> function with
<code>binomial</code> family in R. Optimal codons are determined with a
false discovery rate of 0.001 after multiple testing correction with the
Benjamini-Hochberg procedure (Benjamini and Hochberg 1995) by
default.</p>
</div>
<div class="section level4">
<h4 id="trna-adaptation-index-tai">tRNA Adaptation Index (tAI)<a class="anchor" aria-label="anchor" href="#trna-adaptation-index-tai"></a>
</h4>
<p>tAI quantifies how much the usage of codons in CDS of a gene
resembles the abundance of tRNAs (dos Reis, et al. 2004), which is often
approximated by tRNA gene copy numbers. To determine tAI, the absolute
tRNA adaptiveness value
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>W</mi><mi>i</mi></msub><annotation encoding="application/x-tex">W_i</annotation></semantics></math>
for each codon
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
is defined as</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>W</mi><mi>i</mi></msub><mo>=</mo><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>t</mi><mi>i</mi></msub></munderover><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><msub><mi>s</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><msub><mi>T</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mrow><annotation encoding="application/x-tex">
W_i = \sum_{j=1}^{t_i}{(1 - s_{ij}) T_{ij}}
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>t</mi><mi>i</mi></msub><annotation encoding="application/x-tex">t_i</annotation></semantics></math>
is the number of tRNA isoacceptors recognizing the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>th
codon and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>T</mi><mi>i</mi></msub><annotation encoding="application/x-tex">T_i</annotation></semantics></math>
is the abundance or gene copy number of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>th
tRNA recognizing this codon.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">s_{ij}</annotation></semantics></math>
is a panelty on non-canonical codon–anticodon pairings and differs among
different species (Sabi and Tuller 2014). Cubar uses average
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">s_{ij}</annotation></semantics></math>
values for eukaryotes (Sabi and Tuller 2014) by default. Absolute
adaptiveness values are then normalized by maximum as follows:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>w</mi><mi>i</mi></msub><mo>=</mo><mrow><mo stretchy="true" form="prefix">{</mo><mtable><mtr><mtd columnalign="left" style="text-align: left"><mfrac><msub><mi>W</mi><mi>i</mi></msub><mrow><munder><mo>max</mo><mi>j</mi></munder><msub><mi>W</mi><mi>j</mi></msub></mrow></mfrac><mo>,</mo></mtd><mtd columnalign="left" style="text-align: left"><mrow><mtext mathvariant="normal">if </mtext><mspace width="0.333em"></mspace></mrow><msub><mi>W</mi><mi>i</mi></msub><mo>&gt;</mo><mn>0</mn></mtd></mtr><mtr><mtd columnalign="left" style="text-align: left"><mfrac><mrow><mover><mi>W</mi><mo accent="true">‾</mo></mover><msub><mo stretchy="false" form="prefix">|</mo><mrow><msub><mi>W</mi><mi>j</mi></msub><mo>≠</mo><mn>0</mn></mrow></msub></mrow><mrow><munder><mo>max</mo><mi>j</mi></munder><msub><mi>W</mi><mi>j</mi></msub></mrow></mfrac><mo>,</mo></mtd><mtd columnalign="left" style="text-align: left"><mrow><mtext mathvariant="normal">if </mtext><mspace width="0.333em"></mspace></mrow><msub><mi>W</mi><mi>i</mi></msub><mo>=</mo><mn>0</mn></mtd></mtr></mtable></mrow></mrow><annotation encoding="application/x-tex">
w_i =  \begin{cases} \frac{W_i}{\max_{j}W_j},&amp; \text{if } W_i &gt; 0 \\ \frac{\bar W|_{W_j \neq 0}}{\max_j{W_j}}, &amp; \text{if } W_i=0\end{cases}
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>W</mi><mo accent="true">‾</mo></mover><msub><mo stretchy="false" form="prefix">|</mo><mrow><msub><mi>W</mi><mi>j</mi></msub><mo>≠</mo><mn>0</mn></mrow></msub></mrow><annotation encoding="application/x-tex">\bar W|_{W_j \neq 0}</annotation></semantics></math>
is the geometric mean of non-zero absolute adaptiveness values. Then tAI
of codons in a gene can be calculated as follows, similar to CAI:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">TAI</mtext><mo>=</mo><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><munderover><mo>∏</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>L</mi></munderover><msub><mi>w</mi><mi>k</mi></msub><msup><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo><mfrac><mn>1</mn><mi>L</mi></mfrac></msup><mo>=</mo><mtext mathvariant="normal">exp</mtext><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><mfrac><mn>1</mn><mi>L</mi></mfrac><munderover><mo>∑</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>L</mi></munderover><mo>ln</mo><msub><mi>w</mi><mi>k</mi></msub><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo><mo>=</mo><mtext mathvariant="normal">exp</mtext><mo minsize="2.4" maxsize="2.4" stretchy="false" form="prefix">(</mo><mfrac><mrow><munder><mo>∑</mo><mi>i</mi></munder><munder><mo>∑</mo><mi>j</mi></munder><mrow><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>ln</mo><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mrow><mrow><munder><mo>∑</mo><mi>i</mi></munder><munder><mo>∑</mo><mi>j</mi></munder><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mfrac><mo minsize="2.4" maxsize="2.4" stretchy="false" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">
\text{TAI} = \bigg(\prod_{k=1}^L w_k\bigg)^{\frac{1}{L}} =\text{exp}\bigg(\frac{1}{L}\sum_{k=1}^L \ln w_k\bigg) = \text{exp}\bigg( \frac{\sum_i\sum_j {X_{ij}\ln w_{ij}}}{\sum_i\sum_j X_{ij}} \bigg)
</annotation></semantics></math></p>
</div>
<div class="section level4">
<h4 id="the-mean-codon-stabilization-coefficients-cscg">The mean codon stabilization coefficients (CSCg)<a class="anchor" aria-label="anchor" href="#the-mean-codon-stabilization-coefficients-cscg"></a>
</h4>
<p>CSC of a codon is the Pearson correlation coefficient between the
frequency of that codon and the mRNA half-lives across different genes
(Presnyak, et al. 2015). CSCg is the average codon stabilization
coefficient (CSC) of codons in the CDS of a gene (Carneiro, et
al. 2019):</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">CSCg</mtext><mo>=</mo><mfrac><mn>1</mn><mi>L</mi></mfrac><munderover><mo>∑</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>L</mi></munderover><msub><mtext mathvariant="normal">CSC</mtext><mi>k</mi></msub><mo>=</mo><mfrac><mrow><munder><mo>∑</mo><mi>i</mi></munder><munder><mo>∑</mo><mi>j</mi></munder><mrow><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mtext mathvariant="normal">CSC</mtext><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mrow><mrow><munder><mo>∑</mo><mi>i</mi></munder><mrow><munder><mo>∑</mo><mi>j</mi></munder><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mrow></mfrac></mrow><annotation encoding="application/x-tex">
\text{CSCg} = \frac{1}{L}\sum_{k=1}^L{\text{CSC}_k}= \frac{\sum_i\sum_j {X_{ij} \text{CSC}_{ij}}}{\sum_i{\sum_j{X_{ij}}}}
</annotation></semantics></math></p>
</div>
<div class="section level4">
<h4 id="the-deviation-from-proportionality-dp">The deviation from proportionality (Dp)<a class="anchor" aria-label="anchor" href="#the-deviation-from-proportionality-dp"></a>
</h4>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>D</mi><mi>p</mi></msub><annotation encoding="application/x-tex">D_p</annotation></semantics></math>
measures the departure of the codon usage of an exogenous CDS from the
tRNA pool of the host organism (Chen, et al. 2020; Chen and Yang 2022).
For a codon family
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>n</mi><mi>i</mi></msub><annotation encoding="application/x-tex">n_i</annotation></semantics></math>
synonymous codons
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>n</mi><mi>i</mi></msub><mo>&gt;</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">n_i&gt;1</annotation></semantics></math>),
the fraction of codon
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
among all occurrences of this family in the exogenous CDS is:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mfrac><msub><mi>X</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mi>X</mi><mi>i</mi></msub></mfrac></mrow><annotation encoding="application/x-tex">
Y_{ij} = \frac{X_{ij}}{X_i}
</annotation></semantics></math></p>
<p>The relative tRNA availability of this codon in the codon family in
the host organism is calculated as:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>R</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mfrac><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><msub><mi>n</mi><mi>i</mi></msub></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mfrac></mrow><annotation encoding="application/x-tex">
R_{ij} = \frac{w_{ij}}{\sum_{i=1}^{n_i}{w_{ij}}}
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">w_{ij}</annotation></semantics></math>
can be relative tRNA adaptiveness values as in the calculation of tAI or
the RSCU of host protein-coding genes as in the calculation of CAI. The
discrepancy between codon proportions of exogenous CDS and host tRNA
availability is then calculated as the Euclidean distance:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>D</mi><mi>i</mi></msub><mo>=</mo><msqrt><mrow><munder><mo>∑</mo><mi>j</mi></munder><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>−</mo><msub><mi>R</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></msqrt></mrow><annotation encoding="application/x-tex">
D_i = \sqrt{\sum_j(Y_{ij} - R_{ij})^2}
</annotation></semantics></math></p>
<p>Finally,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>D</mi><mi>p</mi></msub><annotation encoding="application/x-tex">D_p</annotation></semantics></math>
is calculated as the geometric mean of distances for all codon families
with at least two synonymous codons:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>D</mi><mi>p</mi></msub><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><munderover><mo>∏</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>K</mi></munderover><msub><mi>D</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mfrac><mn>1</mn><mi>K</mi></mfrac></msup><mo>=</mo><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><mfrac><mn>1</mn><mi>K</mi></mfrac><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>K</mi></munderover><mrow><mo>ln</mo><msub><mi>D</mi><mi>i</mi></msub></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
D_p = (\prod_{i=1}^K{D_i})^{\frac{1}{K}} = \exp (\frac{1}{K}\sum_{i=1}^K{\ln D_i})
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mo>=</mo><msub><mo>∑</mo><mi>i</mi></msub><mrow><mi>I</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>n</mi><mi>i</mi></msub><mo>&gt;</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mrow><annotation encoding="application/x-tex">K=\sum_{i}{I(n_i&gt;1)}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>I</mi><annotation encoding="application/x-tex">I</annotation></semantics></math>
is the indicator function.</p>
</div>
<div class="section level4">
<h4 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h4>
<ul>
<li>Benjamini Y, Hochberg Y. 1995. Controlling the False Discovery Rate:
A Practical and Powerful Approach to Multiple Testing. Journal of the
Royal Statistical Society: Series B (Methodological) 57:289-300.</li>
<li>Carneiro RL, Requião RD, Rossetto S, Domitrovic T, Palhano FL. 2019.
Codon stabilization coefficient as a metric to gain insights into mRNA
stability and codon bias and their relationships with translation.
Nucleic Acids Res 47:2216-2228.</li>
<li>Chen F, Wu P, Deng S, Zhang H, Hou Y, Hu Z, Zhang J, Chen X, Yang
JR. 2020. Dissimilation of synonymous codon usage bias in virus-host
coevolution due to translational selection. Nat Ecol Evol
4:589-600.</li>
<li>Chen F, Yang JR. 2022. Distinct codon usage bias evolutionary
patterns between weakly and strongly virulent respiratory viruses.
iScience 25:103682.</li>
<li>dos Reis M, Savva R, Wernisch L. 2004. Solving the riddle of codon
usage preferences: a test for translational selection. Nucleic Acids Res
32:5036-5044.</li>
<li>Ikemura T. 1981. Correlation between the abundance of Escherichia
coli transfer RNAs and the occurrence of the respective codons in its
protein genes: a proposal for a synonymous codon choice that is optimal
for the E. coli translational system. J Mol Biol 151:389-409.</li>
<li>Presnyak V, Alhusaini N, Chen YH, Martin S, Morris N, Kline N, Olson
S, Weinberg D, Baker KE, Graveley BR, et al. 2015. Codon optimality is a
major determinant of mRNA stability. Cell 160:1111-1124.</li>
<li>Sabi R, Tuller T. 2014. Modelling the efficiency of codon-tRNA
interactions based on codon usage bias. DNA Res 21:511-526.</li>
<li>Sharp PM, Li WH. 1987. The codon Adaptation Index–a measure of
directional synonymous codon usage bias, and its potential applications.
Nucleic Acids Res 15:1281-1295.</li>
<li>Sun X, Yang Q, Xia X. 2013. An improved implementation of effective
number of codons (nc). Mol Biol Evol 30:191-196.</li>
<li>Wright F. 1990. The ‘effective number of codons’ used in a gene.
Gene 87:23-29.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hong Zhang, Mengyue Liu, Bu Zi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
